File: app/ios/AJ/AJ/AJRuntime.swift
//
//  AJRuntime.swift
//  AJ
//
//  Created by Bruno Fortunato on 17/02/16.
//  Copyright © 2016 Bruno Fortunato. All rights reserved.
//

import UIKit
import JavaScriptCore


open class AJRuntime {
   
    fileprivate var stores = [String: AJStore]()
    fileprivate var plugins = [String: AJPlugin]()
    
    open func getStore(_ type: String) -> AJStore {
        let store = stores[type]
        
        if store != nil {
            return store!
        } else {
            let newStore = AJStore(runtime: self, type: type)
            stores[type] = newStore
            return newStore
        }
    }
    
    open func subscribe(to store: String, owner: AnyObject, subscription: @escaping Subscription) {
        getStore(store).subscribe(owner: owner, subscription: subscription)
        
        NSLog("Subscribed from store: \(store)")
    }
    
    open func unsubscribe(from store: String, owner: AnyObject) {
        getStore(store).unsubscribe(owner: owner)
        
        NSLog("Unsubscribed from store: \(store)")
    }
    
    open func tigger(store: String, data: AJObject) {
        getStore(store).trigger(data)
    }
    
    open func destroy() {
        for store in stores.values {
            store.unsubscribeAll()
        }
        
        stores.removeAll()
    }
    
    open func run(action: String, data: AJObject = AJObject.empty()) -> AJSemaphore {
        fatalError("Not implemented")
    }
    
    open func register(plugin: AJPlugin) {
        NSLog("Plugin registered: \(plugin.name)")
        
        plugins[plugin.name] = plugin
    }
    
    open func get(plugin: String) -> AJPlugin {
        guard let instance = plugins[plugin] else {
            fatalError("Plugin not registered: \(plugin)")
        }
        
        return instance
    }
    
    open func exec(plugin: String, fn: String, data: AJObject) -> AJObject {
        return get(plugin: plugin).exec(fn, data: data)
    }
    
}

File: platforms/android/App/aj/src/main/java/applica/aj/runtime/AJRuntime.java
package applica.aj.runtime;

import android.content.Context;
import android.util.Log;

import java.util.HashMap;
import java.util.Map;

import applica.aj.AJObject;
import applica.aj.Semaphore;
import applica.aj.Store;

/**
 * Created by bimbobruno on 10/03/16.
 */
public abstract class AJRuntime {

    private Map<String, Store> stores = new HashMap<>();
    private Map<String, Plugin> plugins = new HashMap<>();

    protected final Context context;

    public AJRuntime(Context context) {
        this.context = context;
    }

    public Store getStore(String type) {
        if (stores.containsKey(type)) {
            return stores.get(type);
        } else {
            Store store = new Store(this, type);
            stores.put(type, store);

            return store;
        }
    }

    public void subscribe(String store, Object owner, Store.Subscription subscription) {
        getStore(store).subscribe(owner, subscription);
    }

    public void unsubscribe(String store, Object owner) {
        getStore(store).unsubscribe(owner);
    }

    public void trigger(String store, AJObject data) {
        getStore(store).trigger(data);;
    }

    public void registerPlugin(Plugin plugin) {
        assert plugin != null;

        Log.i("AJ", "Registered plugin " + plugin.getName());

        this.plugins.put(plugin.getName(), plugin);
    }

    public Plugin getPlugin(String plugin) {
        if (plugins.containsKey(plugin)) {
            return plugins.get(plugin);
        }

        throw new RuntimeException("Plugin not registered: " + plugin);
    }

    public AJObject exec(final String plugin, final String fn, final AJObject data) {
        Log.i("AJ", String.format("Executing plugin function %s.%s", plugin, fn));

        return getPlugin(plugin).exec(fn, data);
    }

    public void destroy() {
        for (Store store : stores.values()) {
            store.unsubscribeAll();
        }

        stores.clear();
    }

    public Context getContext() {
        return context;
    }

    public Semaphore run(final String action) {
        return run(action, AJObject.empty());
    }

    public abstract Semaphore run(final String action, final AJObject data);


}
